core/builtin.py

```diff
@@ -604,28 +604,51 @@ def Cd(argv, mem, dir_stack):
   return 0
 
 
-WITH_PREFIX = 1
-WITHOUT_PREFIX = 2
+WITH_LINE_NUMBERS = 1
+WITHOUT_LINE_NUMBERS = 2
 SINGLE_LINE = 3
+TILDE_PREFIX_FORM = 1
+LONG_FORM = 2
 
-def _PrintDirStack(dir_stack, style):
+def _PrintDirStack(dir_stack, mem, style, form):
   """Helper for 'dirs'."""
 
-  if style == WITH_PREFIX:
+  if style == WITH_LINE_NUMBERS:
     for i, entry in enumerate(dir_stack.Iter()):
-      print('%2d  %s' % (i, entry))
+      if form == TILDE_PREFIX_FORM:
+        print('%2d  %s' % (i, _TranslateHomePrefix(entry, mem)))
+      elif form == LONG_FORM:
+        print('%2d  %s' % (i, entry))
 
-  elif style == WITHOUT_PREFIX:
+  elif style == WITHOUT_LINE_NUMBERS:
     for entry in dir_stack.Iter():
-      print(entry)
+      if form == TILDE_PREFIX_FORM:
+        print(_TranslateHomePrefix(entry, mem))
+      elif form == LONG_FORM:
+        print(entry)
 
   elif style == SINGLE_LINE:
-    print(' '.join(dir_stack.Iter()))
+    if form == TILDE_PREFIX_FORM:
+      for i, entry in enumerate(dir_stack.Iter()):
+        # Don't add a space after printing the path if it's
+        # the last entry being printed.
+        sys.stdout.write(_TranslateHomePrefix(entry, mem) + \
+          (' ' if (i < dir_stack.Size() - 1) else ''))
+      sys.stdout.write('\n')
+    elif form == LONG_FORM:
+      print(' '.join(dir_stack.Iter()))
 
   sys.stdout.flush()
 
 
-def Pushd(argv, dir_stack):
+def _TranslateHomePrefix(dir_name, mem):
+  val = mem.GetVar('HOME')
+  if val.tag == value_e.Str:
+    return dir_name.replace(val.s, '~', 1)
```

Comment:
I think you have to check if the value starts with the home_dir.  `replace()` will replace it in the middle, which isn't correct.

Reply:
Good point, I can add a check for that.